File: 279.perfect-squares.py
#
# @lc app=leetcode id=279 lang=python3
#
# [279] Perfect Squares
#
import pdb
# @lc code= 
class Solution: #TODO 四平方和定理
    def numSquares(self, n: int) -> int:
        output = [1] #1
        squares = [1]
        for i in range(2,n + 1):
            if i == (len(squares)+1) ** 2:
                squares.append(i)
                output.append(1)
            else:
                min_step =  1 << 32
                for j in range(len(squares)):
                    try:
                        min_step = min(output[i -squares[j]-1 ] , min_step)
                    except:
                        pdb.set_trace()
                output.append(min_step + 1)
        return output[-1]

if __name__ == "__main__":
    print(Solution().numSquares(13))
# @lc code=end

File: golang/279.perfect-squares.go
/*
 * @lc app=leetcode id=279 lang=golang
 *
 * [279] Perfect Squares
 */
package main

import "fmt"

// @lc code=start
func numSquares(n int) int {

	table := make([]int, 101)

	for i := 0; i < 101; i++ {
		table[i] = i * i
	}
	right := 0
	dp := []int{1, 1, 2}
	for i := 3; i < n+1; i++ {
		for table[right] < i {
			right += 1
		}

		if i == table[right] {
			dp = append(dp, 1)
			continue
		}
		min := dp[i-1]
		for j := right; j > 0; j-- {
			if i-table[j] > 0 && min > dp[i-table[j]] {
				min = dp[i-table[j]]
			}
		}
		dp = append(dp, min+1)
	}
	return dp[n]
}

// @lc code=end

func main() {
	fmt.Println(numSquares(12))
}
